cmd/reverse_exporter/main.go

```diff
@@ -38,71 +41,78 @@ func realMain(appConfig AppConfig) int {
 	}
 
 	reverseConfig, err := config.LoadFromFile(appConfig.ConfigFile)
 	if err != nil {
-		log.Errorln("Could not parse configuration file:", err)
-		return 1
+		zap.L().Fatal("Could not parse configuration file", zap.Error(err))
 	}
 
 	// Setup the web UI
 	router := httprouter.New()
 	router = api.NewAPIv1(apiConfig, router)
 
-	log.Debugln("Begin initializing reverse proxy backends")
+	zap.L().Debug("Begin initializing reverse proxy backends")
 	initializedPaths := make(map[string]http.Handler)
 	for _, rp := range reverseConfig.ReverseExporters {
 		if rp.Path == "" {
-			log.Errorln("Blank exporter paths are not allowed.")
-			return 1
+			zap.L().Fatal("Blank exporter paths are not allowed.")
 		}
 
 		if _, found := initializedPaths[rp.Path]; found {
-			log.Errorln("Exporter paths must be unique. %s already exists.", rp.Path)
-			return 1
+			zap.L().Fatal("Exporter paths must be unique", zap.String("already exists", rp.Path))
 		}
 
 		proxyHandler, perr := metricproxy.NewMetricReverseProxy(rp)
 		if perr != nil {
-			log.Errorln("Error initializing reverse proxy for path:", rp.Path)
-			return 1
+			zap.L().Fatal("Error initializing reverse proxy for path", zap.String("path", rp.Path))
 		}
 
 		router.Handler("GET", apiConfig.WrapPath(rp.Path), proxyHandler)
 
 		initializedPaths[rp.Path] = proxyHandler
 	}
-	log.Debugln("Finished initializing reverse proxy backends")
-	log.With("num_reverse_endpoints", len(reverseConfig.ReverseExporters)).Infoln("Initialized backends")
+	zap.L().Debug("Finished initializing reverse proxy backends")
+	zap.L().Info("Initialized backends", zap.Int("num_reverse_endpoints", len(reverseConfig.ReverseExporters)))
 
-	log.Infoln("Starting HTTP server")
-	listenAddrs := strings.Split(appConfig.ListenAddrs, ",")
+	zap.L().Info("Starting HTTP server")
 
-	listeners, listenerErrs, err := multihttp.Listen(listenAddrs, router)
-	defer func() {
-		for _, l := range listeners {
-			if cerr := l.Close(); cerr != nil {
-				log.Errorln("Error while closing listeners (ignored):", cerr)
-			}
-		}
-	}()
+	listener, err := uniListen(appConfig.ListenAddr)
 	if err != nil {
-		log.Errorln("Startup failed for a listener:", err)
-		return 1
+		zap.L().Fatal("Startup failed for a listener", zap.Error(err))
 	}
-	for _, addr := range listenAddrs {
-		log.Infoln("Listening on", addr)
+	zap.L().Info("Listening on", zap.String("addr", appConfig.ListenAddr))
+
+	srv := http.Server{
+		Handler: router,
 	}
 
-	// Setup handlers to catch the listener termination statuses.
-	go func() {
-		for listenerErr := range listenerErrs {
-			log.Errorln("Listener Error:", listenerErr.Error)
+	listenerErrs := make(chan error)
+
+	if appConfig.TLSCert != "" && appConfig.TLSKey != "" {
+		if appConfig.TLSAuthCA != "" {
+			tlsConfig, err := tlsAuthConfig(appConfig.TLSAuthCA)
+			if err != nil {
+				zap.L().Error("Error creating TLS config", zap.Error(err))
+				return 1
+			}
+
+			srv.TLSConfig = tlsConfig
 		}
```

Comment:
there is no case for only TLS-protected endpoint, without mutual TLS auth

Reply:
we do have IP-level auth on the nodes (InputAccept), so there might be such case